packages/flutter/lib/src/scheduler/binding.dart

```diff
@@ -187,6 +197,57 @@ abstract class SchedulerBinding extends BindingBase {
     );
   }
 
+  /// Whether the application is visible, and if so, whether it is currently
+  /// interactive.
+  ///
+  /// This is set by [handleAppLifecycleStateChanged] when the
+  /// [SystemChannels.lifecycle] notification is dispatched.
+  ///
+  /// The preferred way to watch for changes to this value is using
+  /// [WidgetsBindingObserver.didChangeAppLifecycleState].
+  AppLifecycleState get lifecycleState => _lifecycleState;
+  AppLifecycleState _lifecycleState;
+
+  /// Called when the application lifecycle state changes.
+  ///
+  /// Notifies all the observers using
+  /// [WidgetsBindingObserver.didChangeAppLifecycleState].
+  ///
+  /// This method exposes notifications from [SystemChannels.lifecycle].
+  @protected
+  @mustCallSuper
+  void handleAppLifecycleStateChanged(AppLifecycleState state) {
+    _lifecycleState = state;
+    switch (state) {
+      case AppLifecycleState.resumed:
+      case AppLifecycleState.inactive:
+        _setFramesEnabledState(true);
+        break;
+      case AppLifecycleState.paused:
+      case AppLifecycleState.suspending:
+        _setFramesEnabledState(false);
+        break;
+    }
+  }
+
+  Future<String> _handleLifecycleMessage(String message) async {
+    switch (message) {
+      case 'AppLifecycleState.paused':
+        handleAppLifecycleStateChanged(AppLifecycleState.paused);
+        break;
+      case 'AppLifecycleState.resumed':
+        handleAppLifecycleStateChanged(AppLifecycleState.resumed);
+        break;
+      case 'AppLifecycleState.inactive':
+        handleAppLifecycleStateChanged(AppLifecycleState.inactive);
+        break;
+      case 'AppLifecycleState.suspending':
+        handleAppLifecycleStateChanged(AppLifecycleState.suspending);
+        break;
+    }
+    return null;
```

Comment:
Consider making the string-to-AppLifecycleState logic in a private free function to make the control-flow more readable:

```
AppLifecycleState  state = _parseAppLifecycleState(message);
if (state != null)
  handleAp pLifecycleStateChanged(state);
return null;
```

Reply:
Done